File: macOS/OM3Lab/Sources/OM3Lab/OM3Protocol.swift
import Foundation

enum OM3MotionMode: UInt8 {
    case relative = 0x04
    case absolute = 0x05
    case speed = 0x80

    var command: UInt8 {
        switch self {
        case .relative, .absolute:
            return 0x14
        case .speed:
            return 0x0c
        }
    }
}

enum OM3ProtocolError: LocalizedError, Equatable {
    case valueOutOfRange(String)

    var errorDescription: String? {
        switch self {
        case let .valueOutOfRange(field):
            return "\(field) 超出 Int16 范围"
        }
    }
}

enum OM3Protocol {
    static let testStepDegrees = 5
    static let testDurationTenths: UInt8 = 10

    static func crc16<S: Sequence>(_ bytes: S) -> UInt16 where S.Element == UInt8 {
        var crc: UInt16 = 0xdf0c

        for byte in bytes {
            crc ^= UInt16(byte)
            for _ in 0..<8 {
                if crc & 1 == 1 {
                    crc = (crc >> 1) ^ 0x8408
                } else {
                    crc >>= 1
                }
            }
        }

        return crc
    }

    static func rotationMessage(
        yawTenths: Int,
        pitchTenths: Int,
        rollTenths: Int = 0,
        mode: OM3MotionMode = .relative,
        durationTenths: UInt8 = testDurationTenths
    ) throws -> Data {
        let yaw = try int16(yawTenths, field: "yaw")
        let pitch = try int16(pitchTenths, field: "pitch")
        let roll = try int16(rollTenths, field: "roll")

        let header: [UInt8] = [0x55, 0x15, 0x04, 0xa9]
        var body: [UInt8] = [0x02, 0x04, 0x01, 0x00, 0x00, 0x04, mode.command]
        body.append(contentsOf: littleEndianBytes(yaw))
        body.append(contentsOf: littleEndianBytes(roll))
        body.append(contentsOf: littleEndianBytes(pitch))
        body.append(mode.rawValue)
        body.append(durationTenths)

        let checksum = crc16(body)
        return Data(header + body + [UInt8(checksum & 0xff), UInt8(checksum >> 8)])
    }

    static func relativeNudge(yawDegrees: Int, pitchDegrees: Int) throws -> Data {
        try relativeMove(
            yawTenths: yawDegrees * 10,
            pitchTenths: pitchDegrees * 10,
            durationTenths: testDurationTenths
        )
    }

    static func relativeMove(
        yawTenths: Int,
        pitchTenths: Int,
        durationTenths: UInt8
    ) throws -> Data {
        try rotationMessage(
            yawTenths: yawTenths,
            pitchTenths: pitchTenths,
            mode: .relative,
            durationTenths: durationTenths
        )
    }

    static func stopMessage() throws -> Data {
        try rotationMessage(
            yawTenths: 0,
            pitchTenths: 0,
            rollTenths: 0,
            mode: .speed,
            durationTenths: 0
        )
    }

    static func hex(_ data: Data, limit: Int? = nil) -> String {
        let bytes = limit.map { data.prefix($0) } ?? data[...]
        let value = bytes.map { String(format: "%02x", $0) }.joined(separator: " ")
        return limit.map { data.count > $0 ? value + " …" : value } ?? value
    }

    private static func int16(_ value: Int, field: String) throws -> Int16 {
        guard let converted = Int16(exactly: value) else {
            throw OM3ProtocolError.valueOutOfRange(field)
        }
        return converted
    }

    private static func littleEndianBytes(_ value: Int16) -> [UInt8] {
        let bits = UInt16(bitPattern: value)
        return [UInt8(bits & 0xff), UInt8(bits >> 8)]
    }
}

File: app/om3-protocol.ts
export const TEST_STEP_DEGREES = 5;
export const TEST_DURATION_TENTHS = 10;

function int16LE(value: number): number[] {
  const normalized = value & 0xffff;
  return [normalized & 0xff, (normalized >>> 8) & 0xff];
}

export function crc16(bytes: number[]): number {
  let crc = 0xdf0c;

  for (const byte of bytes) {
    crc ^= byte;
    for (let bit = 0; bit < 8; bit += 1) {
      crc = (crc & 1) !== 0 ? (crc >>> 1) ^ 0x8408 : crc >>> 1;
    }
  }

  return crc & 0xffff;
}

export function buildRotationMessage({
  yaw,
  pitch,
  roll = 0,
  mode = "relative",
  time = TEST_DURATION_TENTHS,
}: {
  yaw: number;
  pitch: number;
  roll?: number;
  mode?: "relative" | "absolute" | "speed";
  time?: number;
}): Uint8Array {
  const header = [0x55, 0x15, 0x04, 0xa9];
  const body = [0x02, 0x04, 0x01, 0x00, 0x00, 0x04];
  body.push(mode === "speed" ? 0x0c : 0x14);

  const modeByte = mode === "relative" ? 0x04 : mode === "absolute" ? 0x05 : 0x80;
  const payload = [
    ...int16LE(yaw),
    ...int16LE(roll),
    ...int16LE(pitch),
    modeByte,
    Math.max(0, Math.min(255, time)),
  ];
  const checksum = crc16([...body, ...payload]);

  return Uint8Array.from([
    ...header,
    ...body,
    ...payload,
    checksum & 0xff,
    (checksum >>> 8) & 0xff,
  ]);
}

export function toHex(bytes: Uint8Array): string {
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join(" ");
}
